### PR TITLE
ENH: add grafana monitoring scripts

### PR DESCRIPTION
**Pull request by anish-mudaraddi**
Add scripts that scrape info from openstack that we post to influxdb - and use in our grafana dashboards. 

TODO: 
- ~~Unit Tests~~
- Add docstrings to tests
- CI/CD
-  convert limits_to_influx to only using openstacksdk

updated version of #22

### CODE REVIEW

File: MonitoringTools/usr/local/bin/send_metric_utils.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    """
+    if not data_string:
+        return
+
+    url = f"http://{host}/write?db={db_name}&precision=s"
+    response = requests.post(url, data=data_string, auth=auth)
+    response.raise_for_status()
+
+
+def parse_args(inp_args, description: str = "scrape metrics script") -> Dict:
```
Comment (DavidFair) [inline MonitoringTools/usr/local/bin/send_metric_utils.py:53]: IMO I'd just hardcode the path to somewhere in /etc/

File: MonitoringTools/usr/local/bin/influxdb.conf [hunk 2]
```diff
@@ -0,0 +1,12 @@
+[auth]
+# auth for influxdb
+password=admin
+username=admin
+
+[cloud]
+# requires ~/.config/openstack/clouds.yaml with "prod" cloud account
```
Comment (DavidFair) [inline MonitoringTools/usr/local/bin/influxdb.conf:7]: Nit: this should be in `/etc/openstack` as it's global to the VM

File: MonitoringTools/usr/local/bin/send_metric_utils.py
Comment (DavidFair) [inline MonitoringTools/usr/local/bin/send_metric_utils.py:1]: This file should just be in something like `src`, I want to get this to use the standard pip package layout so we can just build this using CD and pip install it

File: MonitoringTools/usr/local/bin/slottifier_entry.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+                f"Unsupported operand type for +: '{type(self)}' and '{type(other)}'"
+            )
+
+        return SlottifierEntry(
+            *(
+                self_attr + other_attr
+                for self_attr, other_attr in zip(
+                    self.__dict__.values(), other.__dict__.values()
+                )
+            )
```
Comment (DavidFair) [inline MonitoringTools/usr/local/bin/slottifier_entry.py:39]: Nit: do this explicitly, yeah it's a PITA to write but it prevents unexpected surprises and makes it easier to future readers:

```
 slots_available = self.slots_available + other.slots_available
...

return SlottifierEntry(slots_available, ...)
```

File: MonitoringTools/usr/local/bin/send_metric_utils.py (hunk 1 above)
Comment (anish-mudaraddi) [reply MonitoringTools/usr/local/bin/send_metric_utils.py:53]: I'd rather not hardcode.
I'd much rather have paths set and changable in one place - i.e. in Aquilon

File: MonitoringTools/usr/local/bin/send_metric_utils.py
Comment (anish-mudaraddi) [reply MonitoringTools/usr/local/bin/send_metric_utils.py:1]: I'll do that in another PR when we're happy with the code - I'll package-ify it with @meoflynn code #128 (and any other scripts we write in the meantime)

File: MonitoringTools/usr/local/bin/send_metric_utils.py (hunk 1 above)
Comment (anish-mudaraddi) [reply MonitoringTools/usr/local/bin/send_metric_utils.py:53]: This goes double if we want to package-ify it

File: MonitoringTools/usr/local/bin/slottifier.py [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+import openstack
+import sys
+from typing import Dict
+from slottifier_entry import SlottifierEntry
+from send_metric_utils import parse_args, run_scrape
+
+UNKNOWN_GPU_NUM_FLAVORS = []
+
+
+def get_hv_info(hypervisor: Dict, aggregate_info, service_info) -> Dict:
```
Comment (khalford) [inline MonitoringTools/usr/local/bin/slottifier.py:12]: Parameter typing

File: MonitoringTools/usr/local/bin/slottifier.py [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+        hv_info["core_capacity"] = hypervisor["vcpus"]
+        hv_info["mem_capacity"] = hypervisor["memory_size"]
+
+        hv_info["gpu_capacity"] = int(aggregate_info["metadata"].get("gpunum", 0))
+        hv_info["compute_service_status"] = service_info["status"]
+
+    return hv_info
+
+
+def get_flavor_requirements(flavor) -> Dict:
```
Comment (khalford) [inline MonitoringTools/usr/local/bin/slottifier.py:44]: Parameter typing

File: MonitoringTools/usr/local/bin/slottifier.py [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+            f"SlotsAvailable,instance={instance.capitalize()},flavor={flavor}"
+            f" SlotsAvailable={slot_info.slots_available}i"
+            f",maxSlotsAvailable={slot_info.max_gpu_slots_capacity}i"
+            f",usedSlots={slot_info.estimated_gpu_slots_used}i"
+            f",enabledSlots={slot_info.max_gpu_slots_capacity_enabled}i\n"
+        )
+    return data_string
+
+
+def calculate_slots_on_hv(flavor_name, flavor_reqs, hv_info) -> SlottifierEntry:
```
Comment (khalford) [inline MonitoringTools/usr/local/bin/slottifier.py:117]: Parameter typing

**Review by DavidFair — APPROVED** [review]

File: MonitoringTools/usr/local/bin/slottifier.py [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+    all_flavors = {
+        flavor["id"]: flavor for flavor in conn.compute.flavors(get_extra_specs=True)
+    }
+
+    return {
+        "compute_services": list(all_compute_services.values()),
+        "aggregates": list(all_aggregates.values()),
+        "hypervisors": list(all_hypervisors.values()),
+        "flavors": list(all_flavors.values()),
+    }
```
Comment (khalford) [inline MonitoringTools/usr/local/bin/slottifier.py:206]: Can you put this in a variable and return that variable

File: MonitoringTools/usr/local/bin/slottifier.py [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+
+
+def get_all_hv_info_for_aggregate(
+    aggregate, all_compute_services, all_hypervisors
```
Comment (khalford) [inline MonitoringTools/usr/local/bin/slottifier.py:210]: parameter tpying

File: MonitoringTools/usr/local/bin/slottifier.py [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+        hv = next(
+            (
+                hv
+                for hv in all_hypervisors
+                if host_compute_service["host"] == hv["name"]
+            ),
+            None,
+        )
+        if not hv:
+            continue
```
Comment (khalford) [inline MonitoringTools/usr/local/bin/slottifier.py:238]: Can we reduce the amount of comprehension used here?

File: MonitoringTools/usr/local/bin/slottifier.py [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+
+        valid_hvs.append(get_hv_info(hv, aggregate, host_compute_service))
+    return valid_hvs
+
+
+def update_slots(flavors, host_info_list, slots_dict) -> Dict:
```
Comment (khalford) [inline MonitoringTools/usr/local/bin/slottifier.py:244]: Parameter typing

**Review by khalford — CHANGES_REQUESTED** [review]

**Review by khalford — APPROVED** [review]

**Pull request merged**